posts/2024-05-06-check-cfg.md

```diff
@@ -0,0 +1,124 @@
+fn write_shakespeare() {}
+
+#[cfg(windosw)]             // This condition is UNEXPECTED
+                            // it's supposed to be `windows`
+fn win() {}
+```
+
+*`cargo check`*:
+
+![cargo-check](../../../../images/2024-05-06-check-cfg/cargo-check.svg)
+
+## Custom cfgs and build scripts
+
+> In Cargo point-of-view: a custom cfg is one that is neither defined by `rustc` nor by a Cargo feature. Think of `tokio_unstable`, `has_foo`, ... but not `feature = "lasers"`, `unix` or `debug_assertions`
+
+Some crates use custom cfgs that they either expected from the environment (`RUSTFLAGS`or other means) or is enabled by some logic in the crate `build.rs`. For those crates Cargo provides a new instruction: [`cargo::rustc-check-cfg`](https://doc.rust-lang.org/nightly/cargo/reference/build-scripts.html#rustc-check-cfg)[^2] (or `cargo:rustc-check-cfg` for older Cargo version).
+
+[^2]: `cargo::rustc-check-cfg` will start working in Rust 1.80 (or nightly-2024-05-05). From Rust 1.77 to Rust 1.79 *(inclusive)* it is silently ignored. In Rust 1.76 and below a warning is emitted when used without the unstable Cargo flag `-Zcheck-cfg`.
+
+The syntax to use is described in the [rustc book](https://doc.rust-lang.org/nightly/rustc/) section [checking configuration](https://doc.rust-lang.org/nightly/rustc/check-cfg.html), but in a nutshell the basic syntax of `--check-cfg` is:
+
+```
+cfg(name, values("value1", "value2", ..., "valueN"))
+```
+
+Note that every custom cfgs must always be expected, regardless if the cfg is active or not!
+
+### `build.rs` example
+
+`build.rs`:
+```rust
+fn main() {
+    println!("cargo::rustc-check-cfg=cfg(has_foo)");
+    //        ^^^^^^^^^^^^^^^^^^^^^^ new with Cargo 1.80
+    if has_foo() {
+        println!("cargo::rustc-cfg=has_foo");
+    }
+}
+```
+
+> Each `cargo::rustc-cfg` must always have a accompanying **unconditional** `cargo::rustc-check-cfg` directive otherwise you risk having warnings like this: `unexpected cfg condition name: has_foo`.
+
+### Equivalence table
+
+| `cargo::rustc-cfg`      | `cargo::rustc-check-cfg`                       |
+|-------------------------|------------------------------------------------|
+| `foo`                   | `cfg(foo)` or `cfg(foo, values(none()))`       |
+| `foo=""`                | `cfg(foo, values(""))`                         |
+| `foo="bar"`             | `cfg(foo, values("bar"))`                      |
+| `foo="1"` and `foo="2"` | `cfg(foo, values("1", "2"))`                   |
+| `foo="1"` and `bar="2"` | `cfg(foo, values("1"))` and `cfg(bar, values("2"))` |
+| `foo` and `foo="bar"`   | `cfg(foo, values(none(), "bar"))`              |
+
+More details can be found in the [`rustc` book](https://doc.rust-lang.org/nightly/rustc/check-cfg.html).
+
+## Frequently asked questions
+
+### Can it be disabled?
+
+The feature is **always on** and _cannot_ be disabled from Cargo, but like any other lints it can be controlled: `#![warn(unexpected_cfgs)]`.
```

Comment:
```suggestion
For Cargo users, the feature is **always on** and _cannot_ be disabled, but like any other lints it can be controlled: `#![warn(unexpected_cfgs)]`.
```